### PR TITLE
Unify API/Enhance Type-Safety

### PR DESCRIPTION
**Pull request by n1ru4l**
Closes https://github.com/sikanhe/gqtx/pull/49
Closes https://github.com/sikanhe/gqtx/issues/48
Closes https://github.com/sikanhe/gqtx/issues/14
Closes https://github.com/sikanhe/gqtx/issues/37

Hey, @sikanhe I took over #49 and applied some further improvements to the API (optional resolve if the source type has a property that matches the field type) as well as adding some tests for the typescript API and other smaller improvements

### CODE REVIEW

**Comment by sikanhe** [comment]
Interesting,  it looks like we need to make bigger changes to make `field` function to work for both resolver vs no-resolver scenarios? 

- What is the actual problem we are solving with this? Is having a unified API important? Is this actually better than having an explicit function for each use case? This makes the types more complicated while not gaining much. 
- What about `abstractField` ? 
- There are people rely on current API in production - making a big API change like this, unless it is a clear winner, is not preferred.

**Comment by n1ru4l** [comment]
The main problem is actually that the defaultField function right now is not really type-safe. Having both a defaultField and field api does not allow this as afaik, as we cannot cause/raise a typescript error if the defaultField type does not match the Source[fieldName] type.

I missed abstractField, need to add this as well.

afaik gqtx is not adopted that much and the api currently has some flaws. Making the breaking changes now ensures we dont have to do them when the library receives broader adoption.

**Comment by n1ru4l** [comment]
@sikanhe I further adjusted the `abstractField` API and also made the `subscriptionField` API type-safe. Furthermore, I added a TypeScript type test and a jest runtime test to ensure the functionality works as expected.

> What is the actual problem we are solving with this? Is having a unified API important? Is this actually better than having an explicit function for each use case? This makes the types more complicated while not gaining much.

I thought about this a bit more and to me, it seems that we cannot provide a full type-safe API for `field` without complicating the types.

**Comment by n1ru4l** [comment]
Hey @sikanhe, could you have a second look at this?

**Comment by sikanhe** [comment]
@n1ru4l Hey I have been a little bit overwhelmed with my job - I put sometime on my calendar to take a look this this weekend.

**Comment by sikanhe** [comment]
@n1ru4l Hey the PR looks fine - do you think we can move the prettier changes to another PR? It is hard to parse out which lines are the real change

**Comment by n1ru4l** [comment]
@sikanhe yes sure, will ping you once done 😇

**Comment by n1ru4l** [comment]
@sikanhe I reverted all the quote changes that introduced a lot of noise. I can still do more if you want, but I think that should be enough

File: examples/starwars.ts [hunk 1]
```diff
@@ -147,36 +149,40 @@ const episodeEnum = t.enumType({
...
-});
+const characterInterface: Interface<Context, ICharacter | null> =
+  t.interfaceType<ICharacter>({
+    name: 'Character',
+    interfaces: [],
+    fields: () => [
+      t.abstractField('id', t.NonNull(t.ID)),
+      t.abstractField('name', t.NonNull(t.String)),
+      t.abstractField('appearsIn', t.NonNull(t.List(t.NonNull(episodeEnum)))),
+      t.abstractField('friends', characterConnectionType),
```
Comment (sikanhe) [inline examples/starwars.ts:160]: These abstract fields are still using the old style?

**Review by sikanhe — CHANGES_REQUESTED** [review]

File: examples/starwars.ts (hunk 1 above)
Comment (n1ru4l) [reply examples/starwars.ts]: I forgot to update these in the example, fixed it in cebe95c

**Review by sikanhe — APPROVED** [review]

**Comment by n1ru4l** [comment]
@sikanhe Are you fine with me merging this as you approved this?

**Comment by sikanhe** [comment]
@n1ru4l Yes!

**Comment by andrew0** [comment]
Thanks a lot for your effort on this feature - these API changes definitely seem to be better. I'm trying out the canary for this, and I'm running into some strange errors, though. I get errors when I try to use an interface for the source object, but it works fine if I use a type.

For example, this works fine:
```ts
import { createTypesFactory } from 'gqtx';

const t = createTypesFactory();

type TestObject = {
  foo: string;
}

const TestObjectType = t.objectType<TestObject>({
  name: 'foo',
  fields: () => [
    t.field({name: 'foo', type: t.NonNull(t.String)}),
  ],
});
```

If I use an interface, though, I get an error:
```ts
import { createTypesFactory } from 'gqtx';

const t = createTypesFactory();

interface TestObject {
  foo: string;
}

const TestObjectType = t.objectType<TestObject>({
  name: 'foo',
  fields: () => [
    t.field({name: 'foo', type: t.NonNull(t.String)}), // type error on this line
  ],
});
```

```
Type 'Field<undefined, Record<string | number | symbol, unknown>, any, any>' is not assignable to type 'Field<any, TestObject, any, {}>'.
  Types of property 'resolve' are incompatible.
    Type '(src: Record<string | number | symbol, unknown>, args: TOfArgMap<ArgMap<any>>, ctx: undefined, info: GraphQLResolveInfo) => any' is not assignable to type '(src: TestObject, args: TOfArgMap<ArgMap<{}>>, ctx: any, info: GraphQLResolveInfo) => any'.
      Types of parameters 'src' and 'src' are incompatible.
        Type 'TestObject' is not assignable to type 'Record<string | number | symbol, unknown>'.ts(2322)
Argument of type '{ name: "foo"; type: OutputType<undefined, string>; }' is not assignable to parameter of type '{ name: "foo"; type: OutputType<undefined, string>; args?: ArgMap<unknown> | undefined; description?: string | undefined; deprecationReason?: string | undefined; extensions?: unknown; } & ResolvePartialMandatory<...>'.
  Property 'resolve' is missing in type '{ name: "foo"; type: OutputType<undefined, string>; }' but required in type 'ResolvePartialMandatory<Record<string | number | symbol, unknown>, unknown, undefined, string>'.ts(2345)
```

Is this expected?

**Comment by n1ru4l** [comment]
@andrew0 Doed that mean everything is good? We can also add this to the API test cases to ensure it wont break in the future.

**Comment by andrew0** [comment]
@n1ru4l yes, everything is good. The issue was fixed by [this commit](https://github.com/sikanhe/gqtx/pull/52/commits/e6f907e522db45aaaaa878bc727a3ca9c611136c), where `field<TKey extends string, Src extends Record<string | number | symbol, unknown>, Arg, Out>` was changed to `field<TKey extends string, Src, Arg, Out>`. I only ran into the issue because I had accidentally installed an older version.
